### PR TITLE
Preview of new matcher follow-ups :)

### PR DESCRIPTION
**Pull request by ArturGajowy**

### CODE REVIEW

File: presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/MergeLimitWithDistinct.java [hunk 1]
```diff
@@ -48,7 +48,7 @@
     {
         AggregationNode child = captures.get(CHILD);
 
-        if (isDistinct(child)) {
+        if (!isDistinct(child)) {
```
Comment (kokosing) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/MergeLimitWithDistinct.java]: extract this as a separate PR which we could merge on its own

File: presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/MergeLimitWithDistinct.java [hunk 2]
```diff
@@ -35,7 +35,18 @@
     private static final Capture<AggregationNode> CHILD = newCapture();
 
     private static final Pattern<LimitNode> PATTERN = limit()
-            .with(source().matching(aggregation().capturedAs(CHILD)));
+            .with(source().matching(aggregation().capturedAs(CHILD)
```
Comment (kokosing) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/MergeLimitWithDistinct.java]: can you format this a bit differently:

```
with(source()
   .matching(...)
   .matching(...)
   .captured(...)
```
What do you think?

File: presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/MergeLimitWithDistinct.java [hunk 3]
```diff
@@ -35,7 +35,18 @@
...
+                    .matching(aggregation -> isDistinct(aggregation))));
```
Comment (kokosing) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/MergeLimitWithDistinct.java]: You can use a method reference here

File: presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/PushProjectionThroughExchange.java [hunk 4]
```diff
@@ -150,7 +147,7 @@
         return Optional.of(restrictOutputs(context.getIdAllocator(), result, ImmutableSet.copyOf(project.getOutputSymbols())).orElse(result));
     }
 
-    private boolean isSymbolToSymbolProjection(ProjectNode project)
+    private static boolean isSymbolToSymbolProjection(ProjectNode project)
```
Comment (kokosing) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/PushProjectionThroughExchange.java:150]: IMO you can get rid of this methods, just use `ProjectNode::isIdentity`

File: presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/PushTableWriteThroughUnion.java [hunk 5]
```diff
@@ -39,6 +39,12 @@
     private static final Capture<UnionNode> CHILD = newCapture();
 
     private static final Pattern<TableWriterNode> PATTERN = tableWriterNode()
+            // The primary incentive of this optimizer is to increase the parallelism for table
+            // write. For a table with partitioning scheme, parallelism for table writing is
+            // guaranteed regardless of this optimizer. The level of local parallelism will be
+            // determined by LocalExecutionPlanner separately, and shouldn't be a concern of
+            // this optimizer.
```
Comment (kokosing) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/PushTableWriteThroughUnion.java]: maybe put this as a javadoc for class?

File: presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/RemoveEmptyDelete.java [hunk 6]
```diff
@@ -55,12 +53,15 @@
...
 {
-    private static final Capture<ValuesNode> VALUES = newCapture();
-
     private static final Pattern<TableFinishNode> PATTERN = tableFinish()
             .with(source().matching(exchange()
                     .with(onlySource().matching(delete()
-                            .with(source().matching(values().capturedAs(VALUES)))))));
+                            .with(source().matching(emptyValues()))))));
+
+    private static Pattern<ValuesNode> emptyValues()
```
Comment (kokosing) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/RemoveEmptyDelete.java]: I think you can put this to DSL patterns, I would expect to use it more frequently

File: presto-matching/src/test/java/com/facebook/presto/matching/TestMatcher.java [hunk 7]
```diff
@@ -70,11 +70,13 @@ public void propertyMatchers()
         Property<String, Integer> length = Property.property("length", String::length);
         String string = "a";
 
+        assertMatch(aString.with(length.equalTo(1)), string);
```
Comment (kokosing) [inline presto-matching/src/test/java/com/facebook/presto/matching/TestMatcher.java]: So far you could use:
```
assertMatch(aString.with(length.matching(x -> x.equals(1)), string);
```

There is a value in DSL like this, but I do not see much value in specialized `Pattern` I would just reuse `FilterPattern`. Is it because pattern printer?

File: presto-matching/src/main/java/com/facebook/presto/matching/UsageCallSite.java [hunk 8]
```diff
@@ -0,0 +1,88 @@
...
+    private final String fileName;
+    private final int lineNumber;
+
+    public static UsageCallSite get()
+    {
+        StackTraceElement[] stackTrace = new Throwable().getStackTrace();
+        StackTraceElement callSite = stream(stackTrace)
+                .filter(element -> !packageName(element).equals("com.facebook.presto.matching"))
+                .findFirst()
+                .get();
```
Comment (kokosing) [inline presto-matching/src/main/java/com/facebook/presto/matching/UsageCallSite.java:35]: o_O

**Review by kokosing — COMMENTED** [review]
I think the best strategy to create PR to prestodb is to splitt this into several smaller PRs with one per:
 - `HasSource`
 - `FilterPattern` etc
 - `CallSite`
 - `Extractor`

That way in case something got blocked or stuck, the other ones will be able to continue.

File: presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/MergeLimitWithDistinct.java (hunk 2 above)
Comment (ArturGajowy) [reply presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/MergeLimitWithDistinct.java]: I believe the '1 property per row' rule should be used here over the '1 chained method per row' rule.

File: presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/MergeLimitWithDistinct.java (hunk 3 above)
Comment (ArturGajowy) [reply presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/MergeLimitWithDistinct.java]: I tried and changed to this, as this is more readable than `MergeLimitWithDistinct::isDistinct`

File: presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/PushProjectionThroughExchange.java (hunk 4 above)
Comment (ArturGajowy) [reply presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/PushProjectionThroughExchange.java:150]: `ProjectNode::isIdentity` is stricter than this (it also checks expression name = symbol name). I agree that changes not related to pattern matching should be done out of this PR - just like the `isDistinct` change I'm going to split out of this. Feel free to open a PR changing that rule.

File: presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/PushTableWriteThroughUnion.java (hunk 5 above)
Comment (ArturGajowy) [reply presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/PushTableWriteThroughUnion.java]: When I read the comment, I had the impression it does a subtle hint at what's written under it. It's been an implementation comment, I'd prefer to leave it this way.

File: presto-matching/src/test/java/com/facebook/presto/matching/TestMatcher.java (hunk 7 above)
Comment (ArturGajowy) [reply presto-matching/src/test/java/com/facebook/presto/matching/TestMatcher.java]: Several reasons:
1. equals is a common case of a filter pattern one would expect to be available
2. it makes the patterns more readable
3. it allows for better toString at no cost
4. it secures optimization potential
5. it opens the way to building objects from patterns (remember the Builder and Generator interpreters from our recent discussion?)

for these reasons, I strongly believe equalTo patterns should stay.

File: presto-matching/src/main/java/com/facebook/presto/matching/UsageCallSite.java (hunk 8 above)
Comment (ArturGajowy) [reply presto-matching/src/main/java/com/facebook/presto/matching/UsageCallSite.java:35]: ¯\_(ツ)_/¯

Comment (ArturGajowy) [reply presto-matching/src/main/java/com/facebook/presto/matching/UsageCallSite.java:35]: That's the way logging frameworks do that, BTW. (Logback, for example). I'll gladly answer your questions and and arguments if you provide them.

**Comment by kokosing** [comment]
Since there are proper PR created in prestodb, this one is no longer needed.
